Related products helper: rename, cache, correct order

akx opened the pull request:
There's more plumbing than actual fixing in here, really... :)

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/817/
Test PASSed.

suutari-ai commented:
This "fix" is quite BIG change to Shoop Core. Do we really need to introduce a whole new caching system for this? Have you discussed this implementation through with someone?

akx commented:
No, I haven't. Now's as good a time as any, hey?

Also, it's not a very big change, and it's entirely opt-in for things that care about caches.

So far, only two modules in the current master Shoop use `django.core.cache` at all (as you can see from the diff).

The new `shoop.core.cache` module is only a wrapper for `django.core.cache` with better semantics regarding [versioned caches](https://andy.wordpress.com/2008/03/01/cache-with-versioned-tags/).

Using the module should make it very easy to invalidate some cached items (without wholesale clearing the entire cache) -- for instance, when related products are changed, `bump_version("related_products")` will make subsequent attempts of retrieving related products re-cache things.

akx commented:
(Also, considering there haven't been too many other core devs available today, I figure asking for forgiveness rather than permission wasn't a bad policy here...)

akx commented:
(Double also: There's really more docstrings than code in the `shoop.core.cache` module.)

suutari-ai commented:
I bet the MySQL query-bug can be fixed with a lot smaller diff than this and IMHO that fix should be done in a separate PR. When the fix is merged, we should talk about this caching etc. (There were some parts that I'd like to comment about, but don't have time at the moment.)

akx commented:
Well, sure, I'll write the temporary fix, but I'd still like to keep this PR open :)

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/829/
Test PASSed.

Pikkupomo commented:
now that versions change etc, we should check this thru, correct @suutari-ai?

akx commented:
This is rebased on d56dd24954e2921e3d4934a0e56387c84fbf4192 now.

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/993/
Test PASSed.

akx commented:
Rebased on top of 4ffb38aa542346c7886db3d3d8a0cd725026570f

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/1025/
Test PASSed.

akx commented:
Assigning @suutari-ai because

> (There were some parts that I'd like to comment about, but don't have time at the moment.)

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/1280/
Test PASSed.

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/1299/
Test PASSed.

suutari-ai commented on shoop/front/template_helpers/product.py:
```diff
@@ -44,28 +38,46 @@ def is_visible(context, product):
+    :return: List of at most `count` products.
+    :rtype: list[shoop.core.models.Product]
+    """
+    relation_type = map_enum(ProductCrossSellType, relation_type, default=ProductCrossSellType.RELATED)
```
What's the default for here? IMHO it should just raise if API user gives unknown value.

suutari-ai replied:
It would be more useful (for the template developer too) to raise an exception rather than returning possibly incorrect set of products (if e.g. the template developer typoes "related" as "reltaed"). Don't you think?

akx replied:
That's an excellent point I didn't think about.

suutari-ai commented on shoop/utils/enums.py:
```diff
@@ -0,0 +1,36 @@
+    if members is None:
+        members = dict((k.lower(), v) for (k, v) in enum_cls.__members__.items())
+        enum_cls._map_enum_cache_ = members
+    return members.get(six.text_type(value).lower(), default)
```
`default` is not good keyword for a fallback value here, since it's just a way to hide ValueError of unknown enum value.

Is it really needed? I think fallbacking to None is more harmful than raising ValueError.

suutari-ai commented on shoop/utils/models.py:
```diff
@@ -11,6 +11,7 @@
 
 from django.db import models
 from django_countries.fields import Country
+from jinja2.filters import do_batch
```
Using jinja2 filter for this does not feel right. How about using `shoop.utils.iterables.batch` instead?

akx replied:
Sure. The reason it wasn't used in the first place was that, well, it didn't exist back then. (It was only added in 2576caba9614da5008b1b23754722b8c547da65f.)

suutari-ai commented on shoop_tests/utils/test_enum.py:
```diff
@@ -0,0 +1,10 @@
+
+
+def test_map_enum():
+    assert map_enum(ProductMode, 0) is ProductMode.NORMAL
```
What's the point of mapping int to enum with this function? Doesn't it produce unreadable code (because of magic numbers)?

akx replied:
The point is that it is _able_ to do that mapping (which I foresee being useful for, say, import/export situations), but user code probably shouldn't use magic numbers.

suutari-ai replied:
That should be another function IMHO, since it's entirely different use case and having the possibility map from int here allows template developers to write code with magic numbers and we shouldn't make that easy.

suutari-ai commented on shoop_tests/utils/test_enum.py:
```diff
@@ -0,0 +1,10 @@
+    assert map_enum(ProductMode, 0) is ProductMode.NORMAL
+    assert map_enum(ProductMode, "normal") is ProductMode.NORMAL
+    assert map_enum(ProductMode, "Normal") is ProductMode.NORMAL
+    assert map_enum(ProductMode, "shower") is None
```
This last one should raise ValueError (as said in previous comments).

suutari-ai commented on shoop/utils/models.py:
```diff
@@ -49,3 +50,27 @@ def build_or_query(over_fields, term, operator=""):
+
+def get_in_id_order(queryset, id_order, count):
+    """
+    Returns the first `count` identified objects of the given queryset in the given order.
```
Naming of this function is somehow misleading...

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/1391/
Test PASSed.

akx commented:
@suutari-ai:

I taught `map_enum` to not map numbers by default and removed the funny default from `get_related_products` (and added a nicer error message).

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/1417/
Test PASSed.

suutari-ai commented on shoop/utils/enums.py:
```diff
@@ -0,0 +1,64 @@
+    :type default:
+    :param allow_magic_numbers: Allow mapping via "magic numbers". If this is False and
+                                the value is a number, the default `enum_cls(value)` logic
+                                is not attempted.
```
This allow_magic_numbers stuff does not belong to same function IMHO. A separate function would be much clearer, since it's a separate functionality. `map_enum(MyEnum, 42, allow_magic_numbers=True)` vs `map_enum_int(MyEnum, 42)`

Having the int mapping in separate function would also make this one much clearer, since the value would always be string.

akx replied:
The separate function is then just, uh,

```python
def map_enum_int(enum_cls, value):
    return enum_cls(value)
```

I kind of like that `map_enum`, when told so, actively disallows magic numbers.

suutari-ai commented on shoop/utils/models.py:
```diff
@@ -11,6 +11,7 @@
 
 from django.db import models
 from django_countries.fields import Country
+from shoop.utils.iterables import batch
```
grouping

suutari-ai commented on shoop/utils/enums.py:
```diff
@@ -0,0 +1,64 @@
+
+    # Try looking up the value in the mapping.
+
+    retval = members.get(six.text_type(value).lower(), default)
```
I don't like touching private parts of another class (the `_map_enum_cache_` member).

Though... Thinking about this members cache... The only reason it's needed at all is because we can't be sure that if the Enum's members are lower or uppercase (or worst, mixed case). How about just trying it like this:
```python
enum_val = enum_cls.__members__.get(value.upper())
if enum_val is None:
    enum_val = enum_cls.__members__.get(value.lower())
    if enum_val is None:
        for (k, v) in enum_cls.__members__.items():
            if k.lower() == value.lower():
                enum_val = k
                break
```
This would avoid building a cache and would be faster most of the time, since the first or the second will very likely catch all cases. It would be slower for cases where the enum members are mixed case, but that should be very rare.

akx replied:
That looks more complicated though.

My reasoning for building the cache is that -- due to Murphy's law of data imports -- we end up going into the slow path, say, 10,000 times, when some 3rd party has decided that "Related" is a nicer spelling than "RELATED" or "related".

OTOH, maybe it'd be better to just decorate this method with `lru_cache`..?

suutari-ai replied:
You're right it looks more complicated. I think the decoration approach would be nicer.

suutari-ai commented on shoop/front/template_helpers/product.py:
```diff
@@ -44,28 +39,49 @@ def is_visible(context, product):
+    try:
+        relation_type = map_enum(ProductCrossSellType, relation_type)
+    except ValueError as ve:
+        six.raise_from(ValueError("%r is not a valid cross-sell type" % relation_type), ve)
```
`The "Can't map value 'foo' to any member of enum ProductCrossSellType"` error message should be fine, so rather let the exception propagate that mess with it.

akx replied:
Sure.

suutari-ai commented on shoop/front/template_helpers/product.py:
```diff
@@ -44,28 +39,49 @@ def is_visible(context, product):
+    :param product: Source product
+    :type product: shoop.core.models.Product
+    :param relation_type: Relation type (string or ProductCrossSellType value)
+    :type relation_type: str|ProductCrossSellType
```
Why relation_type is not always a string? I don't think there even is any easy way to get the ProductCrossSellType enum value in the template, nor does it make much sense to use that instead of string.

akx replied:
Because this method _does_ accept the actual enum value by virtue of `map_enum`.

suutari-ai replied:
API doc is a contract, not a specification of it's current implementation. It could probably also support any object with pk as product too. ;)

akx replied:
Well, in any case I think it's sane to allow the enum value itself. For instance to support the case

```jinja
{% set relation_type = x.y.z.some_function_from_a_third_party(product) %}
{% for product in get_related_products(product, relation_type=relation_type) %}
...
```

suutari-ai replied:
OK, that's a possible use case (not very like IMHO though), but fine. :)

suutari-ai commented on shoop/themes/classic_gray/templates/classic_gray/shoop/front/macros.jinja:
```diff
@@ -35,7 +35,7 @@
 
 {% macro render_cross_sell_products(product, relation_type="", title="") %}
-    {%- set cross_sell_products = shoop.product.get_product_cross_sells(product, relation_type) %}
+    {%- set cross_sell_products = shoop.product.get_related_products(product, relation_type) %}
```
The default parameter for render_cross_sell_products macro does not seem correct anymore. Shouldn't it be "related"?

akx replied:
Fixed in a future revision, thanks.

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/1418/
Test FAILed.

anders-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://spartacus.aintra:8080//job/shoop-ng-pull-requests/1419/
Test FAILed.

akx commented:
Huh... One would think `from __future__ import absolute_import` was enough to make that `from numbers import Number` work.

suutari-ai commented:
How about just dropping the number test from map_enum and let it always do its job by comparing the value as a string? I.e. remove the part that tries to call the enum constructor with the value. (Also adding the lru_cache while at it.. to make the code cleaner.) That would eliminate the "from numbers import Number" problem. Simple is better than complex.

suutari-ai commented:
This is what I'm after:

```python
from django.utils.lru_cache import lru_cache


RAISE = object()


def map_enum(enum_cls, string, default=RAISE):
    """
    Cast string to value of given enum class.

    ...

    :type enum_cls: enum.EnumMeta
    :type string: str
    :type default: Any
    :rtype: enum.Enum|Any
    """
    value = _map_enum_cached(enum_cls, string)
    if value is None and default is RAISE:
        raise ValueError("Cannot map %r as value of %s" % (string, enum_cls))
    return value if (value is not None) else default


@lru_cache
def _map_enum_cached(enum_cls, string):
    for (key, value) in enum_cls.__members__.items():
        if key.lower() == string.lower():
            return value
    return None
```

Quite simple? Isn't it? (Probably broken too, since it's not tested ;) )

Pikkupomo commented:
Closing this, last comment is months ago. If you want some bits or pieces of this, please make own pr's of those. @akx @suutari-ai
